akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala

```diff
@@ -0,0 +1,372 @@
+import akka.util.JavaDurationConverters.JavaDurationOps
+import akka.util.Timeout
+
+import java.net.URLEncoder
+import java.util.UUID
+import java.util.concurrent.ConcurrentHashMap
+import scala.util.Failure
+import scala.util.Success
+
+/**
+ * INTERNAL API
+ */
+@InternalStableApi
+private[akka] object EventWriterExtension extends ExtensionId[EventWriterExtension] {
+  def createExtension(system: ActorSystem[_]): EventWriterExtension = new EventWriterExtension(system)
+
+  def get(system: ActorSystem[_]): EventWriterExtension = apply(system)
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalStableApi
+private[akka] class EventWriterExtension(system: ActorSystem[_]) extends Extension {
+  private val writersPerJournalId = new ConcurrentHashMap[String, ActorRef[EventWriter.Command]]()
+
+  def writerForJournal(journalId: Option[String]): ActorRef[EventWriter.Command] =
+    writersPerJournalId.computeIfAbsent(
+      journalId.getOrElse(""), { _ =>
+        system.systemActorOf(
+          EventWriter(journalId.getOrElse("")),
+          s"EventWriter-${URLEncoder.encode(journalId.getOrElse("default"), "UTF-8")}")
+      })
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalStableApi
+private[akka] object EventWriter {
+
+  type SeqNr = Long
+  type Pid = String
+
+  sealed trait Command
+  final case class Write(
+      persistenceId: Pid,
+      sequenceNumber: SeqNr,
+      event: Any,
+      metadata: Option[Any],
+      tags: Set[String],
+      replyTo: ActorRef[StatusReply[WriteAck]])
+      extends Command
+  final case class WriteAck(persistenceId: Pid, sequenceNumber: SeqNr)
+
+  private case class MaxSeqNrForPid(persistenceId: Pid, sequenceNumber: SeqNr, originalErrorDesc: String)
+      extends Command
+
+  private def emptyWaitingForWrite = Vector.empty[(PersistentRepr, ActorRef[StatusReply[WriteAck]])]
+  private case class StateForPid(
+      waitingForReply: Map[SeqNr, (PersistentRepr, ActorRef[StatusReply[WriteAck]])],
+      waitingForWrite: Vector[(PersistentRepr, ActorRef[StatusReply[WriteAck]])] = emptyWaitingForWrite,
+      writeErrorHandlingInProgress: Boolean = false,
+      currentTransactionId: Int = 0)
+
+  def apply(journalPluginId: String): Behavior[Command] =
+    Behaviors
+      .supervise(Behaviors.setup[Command] { context =>
+        val journal = Persistence(context.system).journalFor(journalPluginId)
+        context.log.debug("Event writer for journal [{}] starting up", journalPluginId)
+        apply(journal.toTyped)
+      })
+      .onFailure[Exception](SupervisorStrategy.restart)
+
+  def apply(journal: ActorRef[JournalProtocol.Message]): Behavior[Command] =
+    Behaviors
+      .setup[AnyRef] { context =>
+        val maxBatchSize =
+          context.system.settings.config.getInt("akka.persistence.typed.event-writer.max-batch-size")
+        implicit val askTimeout: Timeout =
+          context.system.settings.config.getDuration("akka.persistence.typed.event-writer.ask-timeout").asScala
```

Comment:
maybe more idiomatic to have Settings that is passed in, and read once by the extension